File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        rA, rB = headA, headB
        while rA != rB:
            rA = rA.next if rA is not None else headB
            rB = rB.next if rB is not None else headA
        return rA

File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *rA = headA;
        ListNode *rB = headB;
        while (rA != rB) {
            rA = rA ? rA->next : headB;
            rB = rB ? rB->next : headA; 
        }
        return rA;
    }
};